### PR TITLE
Changed in read for pull#7

### PR DESCRIPTION
**Pull request by rohitsinha54**
@John: A lot changed after you sent the pull request so resolving conflicts was hard. I have done the changed which you have suggested here. 

Please have a look. Thanks.

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -1,17 +1,17 @@
...
 ====================================================================
 
-`Apache Spark <https://spark.apache.org/>`_ is very popular engine to perform in-memory cluster computing for Hadoop. In this guide, you will learn how to run Apache Spark programs with CDAP.
+`Apache Spark <https://spark.apache.org/>`_ is a very popular engine to perform in-memory cluster computing for Apache Hadoop. In this guide, you will learn how to run Apache Spark programs with CDAP.
 
 What You Will Build
 -------------------
 
-You will build a `CDAP application <http://docs.cdap.io/cdap/current/en/dev-guide.html#applications>`_ that exposes REST API to take in web page’s backlinks information and serve out the `PageRank <http://en.wikipedia.org/wiki/PageRank>`_ for the known web pages. You will:
+You will build a `CDAP application <http://docs.cdap.io/cdap/current/en/dev-guide.html#applications>`_ that exposes a REST API to take in web pages' backlinks information and serve out the `PageRank <http://en.wikipedia.org/wiki/PageRank>`_ for the known web pages. You will:
```
Comment (awholegunch) [inline README.rst]: web pages’ backlink information [used curved apostrophe]

File: README.rst [hunk 2]
```diff
@@ -23,19 +23,19 @@ What You Will Need
 Let’s Build It!
 ---------------
 
-Following sections will guide you through building an application from scratch. 
+The following sections will guide you through building an application from scratch. 
 If you are interested in deploying and running the application right away, you 
-can clone its source code from this github repository. In that case feel 
-free to skip the next two sections and jump right to Build & Run section.
+can clone its source code from this GitHub repository. In that case feel
```
Comment (awholegunch) [inline README.rst]: In that case, feel

File: README.rst [hunk 3]
```diff
@@ -23,19 +23,19 @@ What You Will Need
...
-can clone its source code from this github repository. In that case feel 
-free to skip the next two sections and jump right to Build & Run section.
+can clone its source code from this GitHub repository. In that case feel 
+free to skip the next two sections and jump right to the `Build and Run`_ section.
 
 Application Design
 ~~~~~~~~~~~~~~~~~~
 
-Backlinks data is sent to PageRankService over HTTP (e.g. by web crawler as it processes web pages). The service persists the data into backLinks dataset upon receiving. The PageRank for known pages is computed periodically by a PageRankProgram. The program uses backLinks dataset as an input and persists results in pageRanks dataset. 
+Backlinks data is sent to the PageRankService over HTTP (e.g. by a web crawler as it processes web pages). The service persists the data into a backLinks dataset upon receiving it. The PageRank for known pages is computed periodically by a *PageRankProgram*. The program uses the backLinks dataset as an input and persists the results in the pageRanks dataset.
```
Comment (awholegunch) [inline README.rst]: Backlinks data is sent to the _PageRankService_ over HTTP (e.g. by a web crawler as it processes web pages). The service persists the data into a _backLinks_ dataset upon receiving it. The PageRank for known pages is computed periodically by a _PageRankProgram_. The program uses the _backLinks_ dataset as an input and persists the results in the _pageRanks_ dataset.

File: README.rst [hunk 4]
```diff
@@ -55,7 +55,7 @@ The first step is to get the application structure set up.  We will use a standa
 
 The application is identified by the PageRankApp class.  This class extends
```
Comment (awholegunch) [inline README.rst]: the ``PageRankApp`` class.

File: README.rst [hunk 5]
```diff
@@ -55,7 +55,7 @@ The first step is to get the application structure set up.  We will use a standa
 
 The application is identified by the PageRankApp class.  This class extends 
 `AbstractApplication <http://docs.cdap.io/cdap/2.5.0/en/javadocs/co/cask/cdap/api/app/AbstractApplication.html>`_,
-and overrides the configure() method in order to define all of the application components:
+and overrides the configure( ) method to define all of the application components:
```
Comment (awholegunch) [inline README.rst]: ``configure( )``

File: README.rst [hunk 6]
```diff
@@ -73,16 +73,16 @@ and overrides the configure() method in order to define all of the application c
...
       } catch (UnsupportedTypeException e) {
-        throw new RuntimeException("Will never happen: all classes above are supported", e);
+        throw new RuntimeException("Won't happen: all classes above are supported", e);
       }
     }
   }
 
 
-In this example we’ll use Scala to write a Spark program (for example of using Java refer to this `CDAP example <http://docs.cask.co/cdap/current/en/getstarted.html#sparkpagerank-application-example>`_). You’ll need to add scala and maven-scala-plugin as a dependency in your maven `pom.xml <https://github.com/cdap-guides/cdap-spark-guide/blob/develop/pom.xml>`_
+In this example we’ll use Scala to write a Spark program (for example of using Java, refer to the `CDAP example <http://docs.cask.co/cdap/current/en/getstarted.html#sparkpagerank-application-example>`_). You’ll need to add ``scala`` and ``maven-scala-plugin`` as dependencies in your maven `pom.xml <https://github.com/cdap-guides/cdap-spark-guide/blob/develop/pom.xml>`_
```
Comment (awholegunch) [inline README.rst]: for an example of using Java, refer to the `CDAP Java example

File: README.rst [hunk 7]
```diff
@@ -73,16 +73,16 @@ and overrides the configure() method in order to define all of the application c
...
 
 The code below configures Spark in CDAP. This class extends `AbstractSpark <http://docs.cdap.io/cdap/current/en/javadocs/co/cask/cdap/api/spark/AbstractSpark.html>`_
-and overrides the configure() method in order to define all of the components. The setMainClassName method sets Spark Program class.
+and overrides the configure( ) method in order to define all of the components. The setMainClassName method sets the Spark Program class.
```
Comment (awholegunch) [inline README.rst]: ``configure( )``

Comment (awholegunch) [inline README.rst]: The ``setMainClassName`` method specifies the Spark Program class which CDAP will run:

File: README.rst [hunk 8]
```diff
@@ -99,14 +99,14 @@ which CDAP will run:
     }
   }
 
-``BackLinksHandler`` receives backlinks info via POST to ``/backlink``. A valid backlink information is in the form of
-two URLs separated by a whitespace. For example:
+``BackLinksHandler`` receives backlinks info via POST to ``backlink``. A valid backlink information is in the form of
```
Comment (awholegunch) [inline README.rst]: Valid backlink information

File: README.rst [hunk 9]
```diff
@@ -99,14 +99,14 @@ which CDAP will run:
...
+two URLs separated by whitespace. For example:
```
Comment (awholegunch) [inline README.rst]: two URLs separated by whitespace:

File: README.rst [hunk 10]
```diff
@@ -99,14 +99,14 @@ which CDAP will run:
...
 
 .. code::
```
Comment (awholegunch) [inline README.rst]: .. code:: console

File: README.rst [hunk 11]
```diff
@@ -99,14 +99,14 @@ which CDAP will run:
...
 
   http://example.com/page1 http://example.com/page10
   
-BackLinksHandler stores the backlink information in a `ObjectStore Dataset <http://docs.cask.co/cdap/current/en/javadocs/co/cask/cdap/api/dataset/lib/ObjectStore.html>`_ as a String in the format specified above.
+BackLinksHandler stores the backlink information in a `ObjectStore Dataset <http://docs.cask.co/cdap/current/en/javadocs/co/cask/cdap/api/dataset/lib/ObjectStore.html>`_ as a String in the format shown above:
```
Comment (awholegunch) [inline README.rst]: The ``BackLinksHandler`` stores the backlink information in an ``ObjectStore Dataset

File: README.rst [hunk 12]
```diff
@@ -149,8 +149,8 @@ BackLinksHandler stores the backlink information in a `ObjectStore Dataset <http
     }
   }
 
-PageRankProgram Spark program does the actual page rank computation. This code is taken from `Apache Spark's PageRank example <https://github.com/apache/spark/blob/master/examples/src/main/scala/org/apache/spark/examples/SparkPageRank.scala>`_:
-The Spark program stores the computed PageRank in a `ObjectStore Dataset <http://docs.cask.co/cdap/current/en/javadocs/co/cask/cdap/api/dataset/lib/ObjectStore.html>`_ where the key is the URL and the value is the computed PageRank.
+The PageRankProgram Spark program does the actual page rank computation. This code is taken from the `Apache Spark's PageRank example <https://github.com/apache/spark/blob/master/examples/src/main/scala/org/apache/spark/examples/SparkPageRank.scala>`_;
```
Comment (awholegunch) [inline README.rst]: The ``PageRankProgram``

File: README.rst [hunk 13]
```diff
@@ -149,8 +149,8 @@ BackLinksHandler stores the backlink information in a `ObjectStore Dataset <http
...
+the Spark program stores the computed PageRank in a `ObjectStore Dataset <http://docs.cask.co/cdap/current/en/javadocs/co/cask/cdap/api/dataset/lib/ObjectStore.html>`_ where the key is the URL and the value is the computed PageRank:
```
Comment (awholegunch) [inline README.rst:152]: the Spark program stores the computed PageRank in an `ObjectStore Dataset

File: README.rst [hunk 14]
```diff
@@ -182,7 +182,7 @@ The Spark program stores the computed PageRank in a `ObjectStore Dataset <http:/
     }
   }
 
-To serve results out via HTTP let’s add PageRankHandler, which reads PageRank for a given URL from pageRanks dataset:
+To serve results out via HTTP, let’s add a PageRankHandler, which reads the PageRank for a given URL from pageRanks dataset:
```
Comment (awholegunch) [inline README.rst]: a ``PageRankHandler,`` which reads the PageRank for a given URL from the ``pageRanks`` Dataset:

File: README.rst [hunk 15]
```diff
@@ -213,7 +213,7 @@ To serve results out via HTTP let’s add PageRankHandler, which reads PageRank
     }
   }
 
-Build & Run
+Build and Run
 -----------
```
Comment (awholegunch) [inline README.rst]: ------------------------ [add more hyphens!]

File: README.rst [hunk 16]
```diff
@@ -262,10 +262,10 @@ Example output::
 
   0.45521228811700043
 
-Congratulations!  You have now learned how to incorporate Spark data into your CDAP applications.  
+Congratulations!  You have now learned how to incorporate Spark programs into your CDAP applications.  
 Please continue to experiment and extend this sample application.
 
-Share & Discuss
+Share and Discuss
 ---------------
```
Comment (awholegunch) [inline README.rst]: add hyphens....

**Comment by awholegunch** [comment]
Added comments...please review, thanks.

File: README.rst (hunk 6 above)
Comment (rohitsinha54) [inline README.rst]: @awholegunch I don't understand this. Is this right ?

File: README.rst (hunk 13 above)
Comment (rohitsinha54) [inline README.rst:152]: @awholegunch : I don't see a difference here. Can you point it  out ?

File: README.rst (hunk 6 above)
Comment (awholegunch) [inline README.rst]: Yes. I inserted "an" before "example", and "Java" after "CDAP"...if you prefer, we could say:
for an example using Java, refer to the `CDAP SparkPageRank example

File: README.rst [hunk 17]
```diff
@@ -262,11 +261,11 @@ Example output::
...
-Congratulations!  You have now learned how to incorporate Spark data into your CDAP applications.  
+Congratulations!  You have now learned how to incorporate Spark programs into your CDAP applications.  
 Please continue to experiment and extend this sample application.
 
-Share & Discuss
----------------
+Share and Discuss
+------------------
 
 Have a question? Discuss at `CDAP User Mailing List <https://groups.google.com/forum/#!forum/cdap-user>`_
```
Comment (awholegunch) [inline README.rst]: Discuss at the `CDAP User Mailing List. https://groups.google.com/forum/#!forum/cdap-user`_

**Comment by awholegunch** [comment]
In you last commit, you left off a trailing period. Should be: "List. <"

**Comment by awholegunch** [comment]
LGTM.

**Pull request merged**